### PR TITLE
feat: Add the EnforceCache

### PR DESCRIPTION
**Pull request by captt-g**

### CODE REVIEW

**Comment by hsluoyz** [comment]
@xiaobeiswift

**Review by xiaobeiswift — APPROVED** [review]

File: Sources/Casbin/APi/ManagementApi.swift [hunk 1]
```diff
@@ -109,10 +109,10 @@ public extension CoreApi {
         model.getValuesForFieldInPolicy(sec: "p", ptype: ptype, fieldIndex: model.getModel()["p"]?["p"]?.tokens.firstIndex(of: "p_sub") ?? 0)
     }
     func getAllNamedObjects(ptype:String) -> [String] {
-        model.getValuesForFieldInPolicy(sec: "p", ptype: ptype, fieldIndex: model.getModel()["p"]?["p"]?.tokens.firstIndex(of: "p_obj") ?? 1)
+        model.getValuesForFieldInPolicy(sec: "p", ptype: ptype, fieldIndex: model.getModel()["p"]?["p"]?.tokens.firstIndex(of: "p_obj") ??  1)
```
Comment (hsluoyz) [inline Sources/Casbin/APi/ManagementApi.swift:112]: Why?

File: Sources/Casbin/APi/ManagementApi.swift [hunk 2]
```diff
@@ -109,10 +109,10 @@ public extension CoreApi {
...
     }
     func getAllNamedActions(ptype:String) -> [String] {
-        model.getValuesForFieldInPolicy(sec: "p", ptype: ptype, fieldIndex: model.getModel()["p"]?["p"]?.tokens.firstIndex(of: "p_act") ?? 2)
+        model.getValuesForFieldInPolicy(sec: "p", ptype: ptype, fieldIndex: model.getModel()["p"]?["p"]?.tokens.firstIndex(of: "p_act") ??  2)
```
Comment (hsluoyz) [inline Sources/Casbin/APi/ManagementApi.swift:115]: Why?

File: Sources/Casbin/Core/Core.swift [hunk 3]
```diff
@@ -11,7 +11,7 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-
+import NIOConcurrencyHelpers
```
Comment (hsluoyz) [inline Sources/Casbin/Core/Core.swift:14]: No new lines?

File: Sources/Casbin/Core/Core.swift [hunk 4]
```diff
@@ -42,19 +42,28 @@ extension Enforcer {
...
+    
     public struct Core {
         final class Storage {
            
             var eft: Effector
             var watcher:Watcher?
             var rm: RoleManager
             var fm: FunctionMap
-            
+            var locks: Locks
```
Comment (hsluoyz) [inline Sources/Casbin/Core/Core.swift:60]: You should setup a good linter.

File: Sources/Casbin/Enforce+Cache.swift [hunk 5]
```diff
@@ -0,0 +1,105 @@
+//
```
Comment (hsluoyz) [inline Sources/Casbin/Enforce+Cache.swift:1]: No header.

File: Sources/Casbin/Event.swift [hunk 6]
```diff
@@ -68,3 +68,7 @@ func notifyLoggerAndWatcher<T:CoreApi>(eventData:EventData,e: T) {
         w.update(eventData: eventData)
     }
 }
+func clearCache<T:CoreApi>(eventData:EventData,e: T) {
```
Comment (hsluoyz) [inline Sources/Casbin/Event.swift:71]: ?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**